### PR TITLE
Fix Issue 21208 - C/C++ header generation for {string,enum} enums

### PR DESCRIPTION
**Pull request by wilzbach**
Split-off from https://github.com/dlang/dmd/pull/11575

### CODE REVIEW

**Review by MoonlightSentinel — APPROVED** [review]

**Comment by Geod24** [comment]
Could you add a test for:
```D
struct Foo { int i; }
enum MyEnum { A = Foo(42), B = Foo(84) }
```

Otherwise LGTM

**Comment by wilzbach** [comment]
> Could you add a test for:
> 
> ```d
> struct Foo { int i; }
> enum MyEnum { A = Foo(42), B = Foo(84) }
> ```
> 
> Otherwise LGTM

Sure. Added.

**Review by thewilsonator — APPROVED** [review]

**Comment by wilzbach** [comment]
@thewilsonator FYI: I added another test to cover both cases `Foo` being `extern(D)` and `extern(C++)`.

**Comment by thewilsonator** [comment]
I'll leave merging to you then.

File: test/compilable/dtoh_enum.d [hunk 1]
```diff
@@ -86,6 +89,42 @@ END_ENUM_TYPE(const char*, EnumWithImplicitType, ENUMWITHIMPLICITTYPE, enumwithi
...
+
+BEGIN_ENUM(STC, STC, stc)
+    ENUM_KEY(int32_t, a, 1, STC, STC, stc, STC)
+    ENUM_KEY(int32_t, b, 2, STC, STC, stc, STC)
+END_ENUM(STC, STC, stc)
+
+ENUM_CONSTANT(STC, STC_D, (STC)3)
+
+// ignoring non-cpp struct Foo because of linkage
+BEGIN_ENUM_TYPE(Foo, MyEnum, MYENUM, myenum)
```
Comment (wilzbach) [inline test/compilable/dtoh_enum.d:102]: This doesn't make much sense, but I added it to check that there's at least no segfault.

**Pull request merged**